cpp/include/raft/distance/detail/fused_distance_nn/cutlass_base.cuh

```diff
@@ -0,0 +1,161 @@
+{
+  using EpilogueOutputOp = cutlass::epilogue::thread::FusedDistanceNNEpilogueElementwise<
+    DataT,  // ElementC_
+    AccT,   // ElementAccumulator_
+    DataT,  // ElementCompute_
+    AccT,   // ElementZ_
+    OutT,   // ElementT_
+    // 128 / cutlass::sizeof_bits<DataT>::value,
+    1,  // Elements per access 1
+    DistanceFn,
+    CGReduceOpT,
+    ReduceOpT,
+    KVPReduceOpT>;
+  constexpr int batch_count = 1;
+
+  typename EpilogueOutputOp::Params epilog_op_param(
+    dist_op, cg_reduce_op, redOp, pairRedOp, mutexes);
+
+  // Number of pipelines you want to use
+  constexpr int NumStages = 3;
+  // Alignment
+  constexpr int Alignment = VecLen;
+
+  // default initialize problem size with row major inputs
+  auto problem_size = cutlass::gemm::GemmCoord(m, n, k);
+
+  constexpr bool isRowMajor = true;
+
+  using fusedDistanceNNKernel =
+    typename cutlass::gemm::kernel::FusedDistanceNNGemm<DataT,
+                                                        Alignment,
+                                                        DataT,
+                                                        Alignment,
+                                                        AccT,
+                                                        AccT,
+                                                        EpilogueOutputOp,
+                                                        NumStages,  // Number of pipeline stages
+                                                        isRowMajor>::GemmKernel;
+
+  using fusedDistanceNN = cutlass::gemm::device::GemmGrouped<fusedDistanceNNKernel>;
+
+  int num_blocks_per_sm   = fusedDistanceNN::maximum_active_blocks();
+  int num_sms             = raft::getMultiProcessorCount();
+  int full_wave           = num_blocks_per_sm * num_sms;
+  constexpr int mmaShapeM = fusedDistanceNNKernel::Mma::Shape::kM;
+  constexpr int mmaShapeN = fusedDistanceNNKernel::Mma::Shape::kN;
+  int columnTiles         = (problem_size.n() - 1 + mmaShapeN) / mmaShapeN;
+  int rowTiles            = (problem_size.m() - 1 + mmaShapeM) / mmaShapeM;
+
+  int thread_blocks =
+    rowTiles < full_wave ? (columnTiles < full_wave ? columnTiles : full_wave) : rowTiles;
+
+  typename fusedDistanceNN::Arguments arguments{
+    problem_size,
+    batch_count,  // num of problems.
+    thread_blocks,
+    epilog_op_param,
+    x,
+    y,
+    xn,            // C matrix eq vector param, which here is A norm
+    (DataT*)yn,    // this is broadcast vec, which is required to be non-const param
+    dOutput,       // Output distance matrix
+    (int64_t)lda,  // stride A
+    (int64_t)ldb,  // stride B
+    (int64_t)1,    // stride A norm
+    (int64_t)ldd   // stride Output matrix
+  };
+
+  // Using the arguments, query for extra workspace required for matrix multiplication computation
+  size_t workspace_size = fusedDistanceNN::get_workspace_size(arguments);
+  // Allocate workspace memory
+  rmm::device_uvector<uint8_t> workspace(workspace_size, stream);
```

Comment:
but then this will require me to pass the raft handle in this function. is it worth changing the `fusedL2NNImpl` API  for it?

Reply:
Changing `fusedL2NNImpl` would not be a problem, since that is in `detail` namespace, but my suggestion would require changing the public `fusedL2NN` interface as well. Let's skip this for now.